### PR TITLE
Bugfix

### PR DESCRIPTION
**Pull request by ljchang**
Fixes multiple bugs

- removed figure return in all plotting, should remove the double plot issue #241 
- changed keyword `method` to `metric` in Brain_Data and Adjacency `distance` method. fixes issue #313 
- fixes issues with divide by zeros warning in regression - issue #258 
- can now pass in labels with plot_mds to override self.labels.
- fixes issue with label data type in `Adjacency.plot_silhouette`, issue #237

### CODE REVIEW

**Comment by ejolly** [comment]
@ljchang Looks like you have a bug in the divide by zeros fix at line 1011, where you're missing a `e` when indexing the betas array. Would be disastrous as `1.-6 = -5.0`:  

You have:  
`t[stderr > 1.e-3] = b[stderr > 1.-3] / stderr[stderr > 1.e-3]`

Should be:  
`t[stderr > 1.e-3] = b[stderr > 1.e-3] / stderr[stderr > 1.e-3]`

Also most other packages (e.g. numpy, nipy) use `1.e-6` as the convention for too small to matter sigmas as `1.e-3` is still achievable rather than being a rounding error so I would recommend that for better precision.

Happy to do brief code-reviews on any of the bug fixes.

**Comment by ljchang** [comment]
Sounds good. I literally just copied your suggested code.

I’m finishing up grant reviews, but am in the middle of a large push to
clear out some bugs with the toolbox. I’d love to ping you for some code
reviews if your head is back in the game.


On March 20, 2020 at 11:43:46 AM, Eshin Jolly (notifications@github.com)
wrote:

@ljchang <https://github.com/ljchang> Looks like you have a bug in the
divide by zeros fix at line 1011, where you're missing a e when indexing
the betas array. Would be disastrous as 1.-6 = -5.0:

You have:
t[stderr > 1.e-3] = b[stderr > 1.-3] / stderr[stderr > 1.e-3]

Should be:
t[stderr > 1.e-3] = b[stderr > 1.e-3] / stderr[stderr > 1.e-3]

Also most other packages (e.g. numpy, nipy) use 1.e-6 as the convention for
too small to matter sigmas as 1.e-3 is still achievable rather than being a
rounding error so I would recommend that for better precision.

Happy to do brief code-reviews on any of the bug fixes.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/cosanlab/nltools/pull/331#issuecomment-601767137>, or
unsubscribe

**Comment by ljchang** [comment]
fixes issue #320

**Comment by ljchang** [comment]
Fixed #303 can now accomodate numpy arrays and lists

**Comment by ljchang** [comment]
Fixes #269

**Comment by ljchang** [comment]
fixes #264

**Comment by ljchang** [comment]
fixes #309

File: nltools/utils.py [hunk 1]
```diff
@@ -287,15 +287,21 @@ def check_square_numpy_matrix(data):
...
 
-def check_brain_data(data):
+def check_brain_data(data, mask=None):
     '''Check if data is a Brain_Data Instance.'''
     from nltools.data import Brain_Data
 
     if not isinstance(data, Brain_Data):
         if isinstance(data, nib.Nifti1Image):
-            data = Brain_Data(data)
+            if mask is None:
```
Comment (ejolly) [inline nltools/utils.py]: This can be simplified by removing the if statement because `Brain_Data` can already handle a `None` value for mask, so replace with:  

`data = Brain_Data(data, mask=mask)`

**Review by ejolly — CHANGES_REQUESTED** [review]

File: nltools/data/brain_data.py [hunk 2]
```diff
@@ -1205,26 +1205,17 @@ def apply_mask(self, mask):
...
-
+        mask = check_brain_data(mask)
         masked = deepcopy(self)
-        nifti_masker = NiftiMasker(mask_img=mask)
-        masked.data = nifti_masker.fit_transform(self.to_nifti())
-        masked.nifti_masker = nifti_masker
-        if (len(masked.shape()) > 1) & (masked.shape()[0] == 1):
-            masked.data = masked.data.flatten()
+
+        if masked.shape()[0] == mask.shape()[0]:
```
Comment (ejolly) [inline nltools/data/brain_data.py]: I don't think this solution is adequate because line 1208 above (`mask = check_brain_data(mask)`) doesn't explicitly check for what _type_ of data is within the mask. We should either ensure that the mask contains bools only, or change this solution to `masked.data[~mask.data.astype(bool)] = 0`. In fact recent version of numpy will fail with a `TypeError` if you try to invert an array that contains floats using `~`.

File: nltools/data/brain_data.py [hunk 3]
```diff
@@ -1205,26 +1205,17 @@ def apply_mask(self, mask):
...
+            masked.data[~mask.data] = 0
+        else:
+            nifti_masker = NiftiMasker(mask_img=mask)
```
Comment (ejolly) [inline nltools/data/brain_data.py]: Ok so this is converting the original `Brain_Data` instance into whatever space the `mask` is in by creating a `NiftiMasker` out of the `mask`. Should we not be doing the opposite? Using the original `Brain_Data`'s mask to transform the mask we want to apply? Otherwise it's confusing that you could be changing spaces without realizing it.

A second issue is that this will give a different return object compared to how you're dealing with 1d brains on like 1211 above. Above you're _masking out_ non-mask values by setting them to 0 but retaining the original 1d shape. In this solution you could get back fewer voxels (e.g. an ROI with 100 voxels) and thereby get a different return shape. 

The function should be consistent with whether we want to always return back something of the same shape as the original, but with 0'd out voxels, or a subset of the original data shape of the same size as the input mask. I raised this point in issue https://github.com/cosanlab/nltools/issues/304. There are pros and cons for both and our decision could break a lot of existing lab scripts

**Review by ejolly — CHANGES_REQUESTED** [review]

**Comment by ejolly** [comment]
@ljchang Left some code comments for you, most important is how you adjusted `apply_mask` which could lead to some breaking changes in numerous lab scripts. See my point and the discussion I raised previously here: https://github.com/cosanlab/nltools/issues/304

**Comment by ljchang** [comment]
> Also it looks like several functions in `utils.py` and `mask.py` now have `from nltools.data import Brain_Data` _inside_ their function bodies. We should just move them all to the topic of the files like other imports. We're not gaining any speed up, doesn't look like there are any circular import issues (as far as I can tell), and I'm a little concerned this may interact strangely with the `MNI_Template` preference setting, where in some edge cases the specified template gets ignored because of namespace issues.

I'm pretty sure that the only functions that do that are the ones that have circular imports.

**Comment by ljchang** [comment]
> This PR could also be a good time to quickly fix: #104, #272

This PR already fixes #272, either forgot to tag it, or there are multiple open issues with same problem.

I can't replicate your #104, can you give more details?

File: nltools/data/brain_data.py (hunk 3 above)
Comment (ljchang) [reply nltools/data/brain_data.py:1214]: > Ok so this is converting the original `Brain_Data` instance into whatever space the `mask` is in by creating a `NiftiMasker` out of the `mask`. Should we not be doing the opposite? Using the original `Brain_Data`'s mask to transform the mask we want to apply? Otherwise it's confusing that you could be changing spaces without realizing it.
> 

This is the intended use case for this function to convert the data into the mask space. I'm now realizing that the logical indexing approach can return different results, which is problematic. Trying to think about how to solve this.

> A second issue is that this will give a different return object compared to how you're dealing with 1d brains on like 1211 above. Above you're _masking out_ non-mask values by setting them to 0 but retaining the original 1d shape. In this solution you could get back fewer voxels (e.g. an ROI with 100 voxels) and thereby get a different return shape.


> The function should be consistent with whether we want to always return back something of the same shape as the original, but with 0'd out voxels, or a subset of the original data shape of the same size as the input mask. I raised this point in issue #304. There are pros and cons for both and our decision could break a lot of existing lab scripts

Agreed, that this is not desirable for this function. Here I think we want to return the subset of the voxels with the same space as the input mask. The main issue is that we need to use the nifti_masker transform of the mask, otherwise this can create problems transforming back into 3D space.

**Comment by ejolly** [comment]
> > This PR could also be a good time to quickly fix: #104, #272
> 
> This PR already fixes #272, either forgot to tag it, or there are multiple open issues with same problem.

Cool, yup looks like you just forgot to tag it

> I can't replicate your #104, can you give more details?

I added it to the issue, but in short if the `Brain_Data` object that is calling `extract_roi` contains only a single map it's `.data` attribute is typically a 1d numpy array (N voxels) rather than a 2d numpy array with (1 row, N voxels). But `extract_roi` always indexes the `.data` attribute assuming its 2d and fails with an indexing error when its not.

**Comment by ljchang** [comment]
@ejolly: The previous apply_mask function was a mess. Lesson learned to not push code that I didn't test.  I added the boolean indexing option, which now returns an object that is consistent with previous version - hopefully faster now when they are the same size.

**Comment by ljchang** [comment]
Refactored Brain_Data.extract_roi(). Fixes #104. Changed `method` keyword to `metric` for consistency across other functions. Added ability to extract median, and PCA.

**Comment by ljchang** [comment]
Ok, now I'm fixing all tests.  Not going to fix any more issues with this PR.

@ejolly, should we bump this release to 0.3.17 or 0.4.0 once I get all of the tests to pass?

File: nltools/data/brain_data.py (hunk 2 above)
Comment (ljchang) [reply nltools/data/brain_data.py]: Changed the implementation, but now force cast as boolean as suggested.

File: nltools/data/brain_data.py [hunk 4]
```diff
@@ -1229,36 +1267,67 @@ def apply_mask(self, mask):
...
             out: mean within each ROI across images
 
         """
         mask = check_brain_data(mask)
         ma = mask.copy()
-
-        if method != 'mean':
-            raise ValueError('Only mean is currently implemented.')
+            
+        if metric not in metrics:
```
Comment (ejolly) [inline nltools/data/brain_data.py:1287]: `metrics` is undefined in the function body

File: nltools/data/brain_data.py [hunk 5]
```diff
@@ -335,16 +335,54 @@ def shape(self):
...
+
+        '''
 
         out = deepcopy(self)
-        if len(self.shape()) > 1:
-            out.data = np.mean(self.data, axis=0)
-            out.X = pd.DataFrame()
-            out.Y = pd.DataFrame()
-        else:
+        if check_brain_data_is_single(self):
```
Comment (ejolly) [inline nltools/data/brain_data.py:350]: Looks like you're missing an import for `check_brain_data_is_single` at the top of the `brain_data.py` file

**Review by ejolly — COMMENTED** [review]
@ljchang There are a few other issues I'd like to fix this week (e.g. with Design Matrix) so let's bump this PR to 0.3.17.

I'm still not understanding why we're insisting on transforming a `Brain_Data` object into the mask space instead of the mask into `Brain_Data` space in `.apply_mask`?

For example in your latest commit you perform this operation if the number of voxels in the mask don't agree with the number in a `Brain_Data` object. But let's say a mask is in 2mm space  (more voxels) and the brain is in 3mm (fewer). Now the brain gets put into 2mm space without the user knowing because they used a 2mm mask! Outside of nltools you would always resample a mask to the brain-space before applying it. Why don't we do the same? We can still subset the voxels but why not call the `Brain_Data` object's own `NiftiMasker`?

File: nltools/data/brain_data.py (hunk 5 above)
Comment (ljchang) [reply nltools/data/brain_data.py:350]: yeah, a bunch of these small things are getting fixed as I work through getting all of the tests to pass. this one is already fixed.

File: nltools/data/brain_data.py (hunk 4 above)
Comment (ljchang) [reply nltools/data/brain_data.py:1287]: already fixed.

**Comment by ljchang** [comment]
> For example in your latest commit you perform this operation if the number of voxels in the mask don't agree with the number in a `Brain_Data` object. But let's say a mask is in 2mm space (more voxels) and the brain is in 3mm (fewer). Now the brain gets put into 2mm space without the user knowing because they used a 2mm mask! Outside of nltools you would always resample a mask to the brain-space before applying it. Why don't we do the same? We can still subset the voxels but why not call the `Brain_Data` object's own `NiftiMasker`?

This is because the point of the apply mask function is to put the brain into the mask space, not the other way around. If a user desires that functionality, they can transform the mask into the brain_data space first. We could display a warning that the spaces are different, but I really don't want it to automatically change the mask into the brain data space. Users will definitely not expect this and there is no alternative unlike the example I gave, where you resample the mask into brain_space first.

**Comment by ljchang** [comment]
As a compromise, I suppose we could add an optional keyword arg to resample mask into brain space first.

**Comment by ejolly** [comment]
> > For example in your latest commit you perform this operation if the number of voxels in the mask don't agree with the number in a `Brain_Data` object. But let's say a mask is in 2mm space (more voxels) and the brain is in 3mm (fewer). Now the brain gets put into 2mm space without the user knowing because they used a 2mm mask! Outside of nltools you would always resample a mask to the brain-space before applying it. Why don't we do the same? We can still subset the voxels but why not call the `Brain_Data` object's own `NiftiMasker`?
> 
> This is because the point of the apply mask function is to put the brain into the mask space, not the other way around. If a user desires that functionality, they can transform the mask into the brain_data space first. We could display a warning that the spaces are different, but I really don't want it to automatically change the mask into the brain data space. Users will definitely not expect this and there is no alternative unlike the example I gave, where you resample the mask into brain_space first.

I'm confused by this reasoning and I don't think we should handle it this way. It's going to be very unintuitive for end users. Here are 3 reasons why:  

1) Your belief that "users will not expect a mask to change to brain space" doesn't reflects the use cases in the lab and by folks outside the lab (e.g. I'm even looking at old tutorial materials where we've showed folks how to use this). It would break lots of scripts in many projects. Intuitively `brain.apply_mask` is an operation _on an object_ so you expect back a  masked version of _the same kind of object_. Changing resolutions doesn't guarantee that a subset of the _same object_ is returned. 

2) Also it's all too easy (and common) for a user _not_ to check if their desired mask is in the same space as `Brain_Data`; I think think of a dozen examples of this: mask downloaded from a neurosynth atlas, ROI mask obtained from another lab-group, at atlas from a library like FSL or SPM. 

3) From a programming perspective this function has classic "hidden" behavior. The 2 boolean operations when the number of voxels do agree are consistent in their operations - they _subset_ an array, without changing its data. But by creating a new niftimasker, we're fundamentally changing the underlying data. 

Few possible options:
- throw an error if the mask and brain don't have the same resolution
- Add a flag to resample brain-to-mask or mask-to-brain with a default of mask-to-brain
- definitely have a warning

**Comment by ljchang** [comment]
It doesn't break any scripts as it is the exact same method we had before. nothing has changed except for adding the logical indexing when the masks are the same size.

**Comment by ljchang** [comment]
Since we are no longer returning figure handle on plot, @ejolly any idea of how to test if the plot is working other than not crashing when it is generated?

**Comment by ljchang** [comment]
Fixes issue #272

**Comment by ljchang** [comment]
> As a compromise, I suppose we could add an optional keyword arg to resample mask into brain space first.

Done!
